mkosi/__init__.py

```diff
@@ -2299,24 +2299,65 @@ def install_type1(
             f.write("fi\n")
 
 
+def expand_kernel_specifiers(text: str, kver: str, token: str, roothash: str, boot_count: str) -> str:
```

Comment:
This does buy logging unknown specifiers, but it is a lot longer than just replacing the lot? Am I missing some added functionality?

Reply:
I basically took this from the parsing of the primary specifiers.

One case where this could make a difference is when `&&` appears anywhere in the string. For example `&&e` with an entry-token of `kali` would be replaced to `&kali` and then again to `<kver>ali`

Reply:
You're right:
```python
t = '&e-&k-&h+&c'
t.replace("&e", "foo").replace("&k", "6.10-1").replace("&h", "roothash").replace("&c", "15").replace("&&", "&")
# 'foo-6.10-1-roothash+15'
expand_kernel_specifiers(t, "6.10-1", "foo", "roothash", "15")
# 'foo-6.10-1-roothash+15'
s = "&&close-&&m-&g&e-&k-&h+&c+&&+&f"
s.replace("&e", "foo").replace("&k", "6.10-1").replace("&h", "roothash").replace("&c", "15").replace("&&", "&")
# '&15lose-&m-&gfoo-6.10-1-roothash+15+&+&f'
expand_kernel_specifiers(s, "6.10-1", "foo", "roothash", "15")
WARNING:root:Unknown specifier '&g' found in &&close-&&m-&g&e-&k-&h+&c+&&+&f, ignoring
WARNING:root:Unknown specifier '&f' found in &&close-&&m-&g&e-&k-&h+&c+&&+&f, ignoring
# '&close-&m-foo-6.10-1-roothash+15+&+'
```
Dropping the unknown specifiers outright also seems a bit like a footgun.  

How about this? I think this should generally work and it's a lot shorter
```python
def expand_kernel_specifiers(text: str, kver: str, token: str, roothash: str, boot_count: str) -> str:
     specifiers = {
         "&": "&",
         "e": token,
         "k": kver,
         "h": roothash,
         "c": boot_count
     }
     return re.sub(r"(?<!&)&(?P<specifier>[&ekhc])", lambda m: specifiers.get(m.group("specifier"), ""), text)
```

Reply:
I'm not too sure about using regex. Ontop of being basically illegible to basically anyone that doesn't look it up, this would also need to keep the list of accepted specifiers in sync with the dictionary.

Reply:
I think a better solution would be to add comments, because even with comments, it would probably still be shorter than what we have so far.

For completeness the regex version of the current code
```python
def expand_kernel_specifiers(text: str, kver: str, token: str, roothash: str, boot_count: str) -> str:
    specifiers = {
        "&": "&",
        "e": token,
        "k": kver,
        "h": roothash,
        "c": boot_count
    }

    def replacer(match: re.Match) -> str
        m = match.group("specifier")
        if specifier := specifiers.get(m):
            return specifier
        logging.warning(f"Unknown specifier '&{m}' found in {text}, ignoring")
        return ""

    return re.sub(rf"&(?P<specifier>[&a-zA-Z])", replacer, text)
```

Reply:
CI does not seem to like `def replacer(match: re.Match) -> str:` and I am unsure what exactly it's referencing/wants fixed.